File: bot-example/java/src/main/java/org/example/Strategy.java
package org.example;

import java.util.List;

public class Strategy {
    private Api.Options gameOptions;

    public void onJoinGame(Api.Options gameOptions) {
        this.gameOptions = gameOptions;
    }

    public Action getAction(Api.GameState gameState) {
        for (Api.GameState.Player player : gameState.players) {
            if (player.isMe) {
                Api.GameState.Drone drone = player.drone;
                Api.Options.CellPos nextCellPoint = gameOptions.maze.checkpoints.get(drone.nextCheckpointIndex);
                List<Api.Options.CellPos> path = PathFinding.solve(
                        gameOptions.maze,
                        Utils.posToGrid(gameOptions, drone.pos),
                        nextCellPoint
                );

                if (path.size() > 1) {
                    Api.Options.CellPos goal = path.get(1);
                    Api.Vec2 nextPoint = Utils.gridToPos(gameOptions, goal);
                    Api.Vec2 force = Utils.normalizeVector(Utils.calculateDirectionVector(drone.pos, nextPoint));
                    return new Action(force.x * 500, force.y * 500, 0f);
                }
            }
        }

        return new Action(0f, 0f, 0f);
    }
}

File: bot-example/csharp/Strategy.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using CSharp.Models;

namespace CSharp
{
    public class Strategy
    {
        private readonly Options _options;

        public Strategy(Options options)
        {
            _options = options;
        }

        public Models.Action GetAction(GameState gameState)
        {
            foreach (var player in gameState.Players)
            {
                if (player.IsMe)
                {
                    var drone = player.Drone;
                    var nextCellPoint = _options.Maze.Checkpoints[drone.NextCheckpointIndex];

                    var path = PathFinding.Solve(
                        _options.Maze,
                        _options.GetCellPosition(drone.Position),
                        nextCellPoint
                    );

                    if (path.Count > 1)
                    {
                        var goal = path[1];

                        var nextPoint = _options
                            .GetPosition(goal)
                            .Subtract(drone.Position)
                            .Normalize();

                        return new Models.Action(nextPoint.X * 500, nextPoint.Y * 500, 0f);
                    }
                }
            }

            return new Models.Action(0f, 0f, 0f);
        }
    }
}
